### ISSUE TITLE
DefaultAzureCredential chain crashing if ImdsCredential _get_token call hits proxy server that returns HTML

### ISSUE THREAD

**Issue by Chuck1823** [Azure/azure-sdk-for-python#35166]
- **Package Name**: azure-identity
- **Package Version**: 1.15.0
- **Operating System**: MacOS Sonoma 14.3.1
- **Python Version**: 3.9.10

**Describe the bug**

DefaultAzureCredential, which inherits from ChainedTokenCredential, loops through different auth mechanisms in order to get a token to authenticate the user. One such authentication scheme from the chain is the ManagedIdentityCredential which in turn uses ImdsCredential. ImdsCredential tries to hit an endpoint. In my case, I am sitting behind a corporate proxy which is throwing an HTTP error 504 with an HTML-based error message. The ImdsCredential code does not check for that and cannto handle HTML so it throws a ClientAuthenticationError instead of a CredentialUnavailableError and halts the procession of ChainedTokenCredential, preventing it from hitting the other authentication mechanisms that I am targetting.

**To Reproduce**
1. Operate in an environment that requires to go through a proxy or some API gateway before hitting any endpoint that is in the public internet.
2. Go through the ChainedTokenCredential authentication scheme and ensure that the proxy server or gateway returns a 504 with an HTML-based message
3. Watch ChainedTokenCredential crash and burn instead of moving on to the next authentication mechanism

**Expected behavior**

The ImdsCredential class should handle the 504 and the HTML error message, throw a CredentialUnavailableError exception and move on to the next auth mechanism from ChainedTokenCredential

**Screenshots**

**Additional context**

**Comment by xiangyan99** [comment]
Thanks for the feedback, we’ll investigate asap.

**Comment by Chuck1823** [comment]
> Thanks for the feedback, we’ll investigate asap.

@xiangyan99 I want this to be my first ever open source contribution!!! I'm working on a PR to fix it and will link this issue to the PR shortly. Thanks for the lightning fast reply!

**Comment by xiangyan99** [comment]
Thank you.

We are under discussions what is the right behavior for it.

In some scenarios, 504 is valid response for IMDS and we don't want to silently swallow the error and try next credential.

As a work around for your case, I would suggest you disabling managed identity if you run your app behind a proxy.

**Comment by Chuck1823** [comment]
Yes, that is what I did (a few months back). My suggestion, as you will see in the PR, is to wrap the _check_forbidden_response (where it ultimately fails) call into a try/catch that handles exceptions thrown if the parsing of the Http error message fails. That way, if a true 504 with a valid, non-html response is sent back, the program will not fail silently as expected

**Comment by Chuck1823** [comment]
I just opened up the PR #35170

**Comment by pvaneck** [comment]
Hey @Chuck1823, to run tests locally, you can check the [CONTRIBUTING.md file](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md#whl-environment). Typically, to run the unit tests locally, from the azure-identity directory, I'd just run `tox run -e whl -c ../../../eng/tox/tox.ini --root .`

Also, can you provide a stack trace of the error you were encountering?
